### ISSUE TITLE
Comments on "Reading QR codes with an MRTK2 Extension Service"

### ISSUE THREAD

**Issue by LocalJoost** [LocalJoost/BlogComments#371]
Original article: https://localjoost.github.io/Reading-QR-codes-with-an-MRTK2-Extension-Service/

**Comment by ephynes** [comment]
Is it possible using this code to have access to the raw camera frames?

I want to apply OpenCV filters to the frames to augment them.

**Comment by LocalJoost** [comment]
There is something like [Research Mode](https://docs.microsoft.com/en-us/windows/mixed-reality/develop/platform-capabilities-and-apis/research-mode), which will give you access to the raw camera feed. Not sure if that is what you need. I have no experience with it.

**Comment by ephynes** [comment]
Yes I have used it. My problem is that now I also need to track the users eye gaze. I don't think the research mode API supports eye tracking, so I am trying to find a way to access raw video feed within MRTK. There are some projects that demonstrate how to do this but they have some deprecated code that I cannot find an up to date alternative for.

In your code I cannot see any reference to the camera input.
Thanks for your prompt reply.

**Comment by ephynes** [comment]
I got this repo working on HL2: https://github.com/camnewnham/HoloLensCameraStream

It allows for access to raw pv camera frames with MRTK 2.2.

**Comment by Ryxuu** [comment]
Dear Joost,

I deployed your solution to HoloLens 2. I use the same version of Unity, MRTK and MRTK QR package. Unity loads the project correctly and builds it. I can deploy it with Visual Studio 2019 without errors. When the app starts I give it all the permissions it requires. The Debug logger panel appears and reads "QR tracker initialized" and "Enabled tracking" and if I show a QR code it crashes with the following debug output:

Exception thrown at 0x00007FFCD59CA71C in Networking.exe: Microsoft C++ exception: Il2CppExceptionWrapper at memory location 0x0000006CF0EFDAC0.
Exception thrown at 0x00007FFCD59CA71C in Networking.exe: Microsoft C++ exception: Il2CppExceptionWrapper at memory location 0x0000006CF0EFF830.
Exception thrown at 0x00007FFCD59CA71C (KernelBase.dll) in Networking.exe: WinRT originate error - 0x80004003 : 'Invalid pointer'.
'Networking.exe' (Win32): Loaded 'C:\Windows\system32\msxml6.dll'.
Exception thrown at 0x00007FFCD59CA71C in Networking.exe: Microsoft C++ exception: winrt::hresult_error at memory location 0x0000006CF0EFF920.

What could be the problem? Can you help me?

If you need more info please let me know.

Thank you!

Best,
Abraham

**Comment by loleklel** [comment]
I can build the project for hololens 2 without any errors but the QR recognition does not work. I tried both scenes in your project without success. According to debug logger: QR tracker is initialized, tracking enabled, and initializing QR tracker attempt 1 is shown. I tried differents codes but nothing happens.

thanks for reply

**Comment by LocalJoost** [comment]
@loleklel please provide context:

1. What version of Unity do you use?
2. What version of Visual Studio do you use?
3. What QR codes do you use?
4. How big QR codes to you use?
5. What branch did you clone and WHEN?
6. What settings did you use in Visual Studio when building the app (Master/ARM64?)

And please read through this thread. https://github.com/LocalJoost/QRCodeService/issues/2#issuecomment-797622719. It works, I am sure of it, it's always 'some setting'

**Comment by loleklel** [comment]
well, it was a fast response! 
1. Unity 2019.4.22f1
2. Visual Studio 2019 16.9.0
3. tried different QR codes from Microsoft demo page, your and big sizes as well
4. 10x10 cm or smaller or bigger (doesn't really matter because in other apps they are working well)
5. main branche
6. I build with unity (Release, ARM64)

**Comment by loleklel** [comment]
well, I just tried to build with the master and copied PDB files, and apparently, it works!

**Comment by LocalJoost** [comment]
I see. Well I always build with Master, unless I want to debug - then I use debug. But never release. Glad it worked for you!

**Comment by loleklel** [comment]
> I see. Well I always build with Master, unless I want to debug - then I use debug. But never release. Glad it worked for you!

thanks!

**Comment by nqwifly** [comment]
Do you think  QR for mixed reality could work for Android/iOS?

**Comment by LocalJoost** [comment]
I think it won't. The packages I used seem to be pretty much Windows (Holographic) specific

**Comment by melixor** [comment]
Hi i'm tring to build the application on Unity 2020.3.30f1 and i have a problem. 
It seems that legacy XR pipeline, is deprecated in Unity 2019 and removed in Unity 2020. 
So i receive an error like thate: 
"Unable to register IMixedRealityInputDeviceManager data provider (Windows Mixed Reality Device Manager) because the value of concreteType is null." 
etc.

Do you have tested it on this Unity version?

**Comment by LocalJoost** [comment]
Hi @melixor, if you read [this article ](https://localjoost.github.io/Upgrading-reading-and-positioning-QR-codes-with-HoloLens-2-to-Unity-2020-+-OpenXR-plugin/) from my blog and pull the [OpenXR branch](https://github.com/LocalJoost/QRCodeService/tree/openxr) from this very project, I think you will do just fine

**Comment by ARVR-damir** [comment]
Hi. I am starting Unity developer. I am trying to understand your example. I start the project and Unity gives an error: "Unable to find IQRCodeTrackingService service.". Please help me solve this problem. I will be very thankful.

**Comment by LocalJoost** [comment]
@ARVR-damir Listen it does not work in the Unity editor. It only work on HoloLens itself. Please read https://localjoost.github.io/Positioning-QR-codes-in-space-with-HoloLens-2-building-a-'poor-man's-Vuforia'/#conclusion first. And if you are really starting with Unity, this is quite complicated stuff. There are a lot of things you must get right. Read the "Please note" section carefully

**Comment by ARVR-damir** [comment]
@LocalJoost thanks a lot!!!

**Comment by h4ergo** [comment]
Hello, thanks for the tutorial! 
I have a question, i am trying to redo all the steps in your article but when i run the app i get always: **QR Tracker could not be initialized: NotDeclaredByApp ... attempt 1**, any suggestion?

Thanks! ;)

**Comment by LocalJoost** [comment]
@h4ergo you have most likely forgotten to set the WebCam capability.

**Comment by Diesellisel** [comment]
Hello, first of thank you for the turorial it is really helpfull! I'm new to this and i tried to run your projekt on the hololens 2, but it dont react with the qr code. I get the debug message in unity "QR tracking not supoorted" Can you help me?
Thank you!

**Comment by LocalJoost** [comment]
Hello @Diesellisel , as you can read above: you are not the first one to report is doesn't work, but in the end, it always does. So. 
- Have you built and deployed for Master (release and debug WON'T WORK)
- Do you run the app _at the HoloLens_ and not from Visual Studio or inside Unity?
- Have you checked the WebCam capability? In Unity AND in Visual Studio?

Read https://localjoost.github.io/Positioning-QR-codes-in-space-with-HoloLens-2-building-a-'poor-man's-Vuforia'/#conclusion and then especially the "please note" part

**Comment by GabrielleSamson** [comment]
Hi ! I'm new with Unity and Hololens and I'm trying to make your project work step by step on my own existing one (I didn't dowload your project). I managed to dowload Microsoft.MixedReality.QR but I don't understand how it works. You don't write something like "using Microsoft.MixedReality.QR" anywhere to use the package (and this line doesn't work because "QR" is apparently not known), and in QRTrackerController for example, the types like QRInfo, IQRCodeTrackingService... are not recognized. I feel like it's a problem coming from me not using the package in the right way but I don't know how to do so. Do you know how can I solve my problem ? 
Thank you !

**Comment by LocalJoost** [comment]
Hello @GabrielleSamson If I were you, I would first read https://localjoost.github.io/Positioning-QR-codes-in-space-with-HoloLens-2-building-a-'poor-man's-Vuforia'/#conclusion and the "Please Note" part
